stan/math/rev/fun/inverse.hpp

```diff
@@ -75,16 +21,23 @@ class inverse_vari : public vari {
  * size zero).
  * @throw std::invalid_argument if the matrix is not square.
  */
-inline matrix_v inverse(const matrix_v &m) {
+template <typename T, require_eigen_vt<is_var, T>* = nullptr>
+inline Eigen::Matrix<var, Eigen::Dynamic, Eigen::Dynamic> inverse(const T& m) {
   check_square("inverse", "m", m);
   if (m.size() == 0) {
     return {};
   }
 
-  matrix_v res(m.rows(), m.cols());
-  internal::inverse_vari *baseVari = new internal::inverse_vari(m);
-  res.vi() = Eigen::Map<matrix_vi>(baseVari->vari_ref_A_inv_, res.rows(),
-                                   res.cols());
+  arena_matrix<Eigen::Matrix<var, Eigen::Dynamic, Eigen::Dynamic>> arena_m = m;
```

Comment:
Mostly asking because it looks like in most places we than access these things by `.val()` or `.adj()`. I think it's just the cost calculation of whether it's better to allocate N vars and have non-contiguous access on the `val()` and `adj()` or make N*2 double allocations + 1 vari allocation and have contiguous access on the `val()` and `adj()`

Reply:
Well at the end we have to update adjoints of the `m`. I think right now we do not have an operation that would propagate adjoints form `var_value<matrix>` to `Matrix<var>`. Even if we did it would do exactly what is done here. So doing it directly is preferable.

Reply:
Once we start supporting `var_value` we need something here that will support both option, such as
```suggestion
  arena_t<T> arena_m = m;
```

Reply:
I'm not sure we need it. If the input is a `Matrix<var, R, C>` type, copying to a `var_value<Matrix>` type only saves us time if the adjoints and vals are accessed multiple times. Adjoints are basically only accessed once ever in a custom reverse pass implementation. If values are used more than once, they can be copied out separately.